A29-xds-tls-security.md

```diff
@@ -0,0 +1,340 @@
+A29: xDS-Based Security for gRPC Clients and Servers
+----
+* Author(s): Sanjay M. Pujare
+* Approver: a11r
+* Status: Draft
+* Implemented in: C-core, Java, and Go (planned)
+* Last updated: 2020-06-04
+* Discussion at: https://groups.google.com/g/grpc-io/c/4IwVLPeTAe4
+
+## Abstract
+
+This proposal describes adding transport security via TLS or mTLS to xDS-orchestrated
+gRPC connections. TLS/mTLS adds encryption and authentication capabilities to the
+gRPC connections. Other aspects of security \- such as RBAC \- are not covered.
+
+This is an incremental feature over existing
+[xDS support](https://github.com/grpc/proposal/blob/master/A27-xds-global-load-balancing.md)
+in gRPC to add transport security capabilities offered by xDS and present in Envoy.
+gRPC will be adding more xDS capabilities over a period of time to match Envoy's
+as part of the efforts to provide a proxyless alternative to Envoy.
+
+## Background
+
+The xDS API supports both TLS termination of incoming connections in listeners and
+TLS origination of outbound connections in clients. The TLS configuration on the listener side
+is provided in the
+[Listener](https://github.com/envoyproxy/envoy/blob/master/api/envoy/config/listener/v3/listener.proto)
+resource and on the client side in the [Cluster](https://github.com/envoyproxy/envoy/blob/master/api/envoy/config/cluster/v3/cluster.proto) resource.
+This proposal covers supporting a subset of these TLS configurations to enable
+gRPC traffic to be automatically secured by the xDS control plane thereby eliminating
+the overhead of manually provisioning and managing TLS certificates by gRPC developers.
+
+We leverage the xDS API's TLS configuration capabilities to achieve this and extend
+gRPC's channel and server credentials concept to enable users to take advantage of
+the xDS control plane. The proposal is based on version 3 of the xDS API but
+the support will be upgraded to match newer xDS API versions. The proposal depends on
+the
+[Certificate Provider](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/transport_sockets/tls/v3/tls.proto#extensions-transport-sockets-tls-v3-commontlscontext-certificateprovider)
+additions made to xDS API v3 to support agentless transport security.
+
+### Related Proposals
+[A27: xDS-Based Global Load Balancing](https://raw.githubusercontent.com/grpc/proposal/master/A27-xds-global-load-balancing.md)
+
+## Proposal
+
+### Usage
+
+A gRPC developer explicitly signals the use of this feature by using a new type of
+credentials on the client side and server side.
+
+#### Client Side Usage
+
+A prerequisite is to use xDS on the client side as described
+[here](https://github.com/grpc/proposal/blob/master/A27-xds-global-load-balancing.md#grpc-client-architecture).
+The gRPC developer also needs to use a new type of (client-side) credentials called
+`XdsCredentials` to use the feature on the client side. An example of this in C++ is:
```

Comment:
This seems to drift from how Envoy behaves in unexpected ways. This is all from the perspective of a user, and not someone familiar with gRPC internals, but this seems un-intuitive to me.

Today, when writing an XDS control plane, the control plane is the source of truth and declares *all* behavior for the client. Aside from mismatched versions (sending new fields to old Envoys), if I send the same config to two Envoys I would expect them to both start using that configuration.

However, by requiring `XdsCredentials`, I believe that is no longer the case. The control plane may send the gRPC client some config to use mTLS, and its dropped on the floor. I think at the very least this could use some indication (perhaps client_features?) to the server that its going to ignore the TLS settings. As a control plane developer, its already confusing enough which fields gRPC supports - if its selectively supporting them it sounds like a nightmare to develop around, and certainly a major challenge to just drop an existing XDS server in place (rather than making a custom gRPC XDS server).

One example of where this could blow up - Istio's XDS server automatically determines whether to tell clients/servers to use mTLS based on its global view of the world. If well tell all connected clients to use mTLS, but half of them decide to ignore it, we are going to have problems.

I also feel like this has a dangerous precedent. Why is TLS the only setting that can override the XDS response? Will we soon need add a `grpc::XdsRetryOptions`, a `grpc::XdsLoadBalancingPolicy`, etc that overrides the XDS response and applies some local rules instead? If so, we may suffer the same issues mentioned above. If not, it feels inconsistent to me.

Finally, by adding this requirement, there are artificial barriers to adoption, as a user cannot just drop in the xds:/// scheme, but rather needs more code changes. This is discussed extensively in https://github.com/grpc/grpc-go/issues/4123 though, so it may make sense to continue the discussion there instead of splitting the conversation.

Reply:
Unfortunately, we can't do that with breaking the security guarantees of our existing API.  The gRPC API requires the application to explicitly indicate the type of credentials to use, and they have an expectation that we will use the type of credentials that they told us to use.  We cannot simply decide to ignore what the user asked us to do and instead use what xDS tells us to do, because (e.g.) if they told us to use TLS and xDS told us to use plaintext, we would wind up sending their data unencrypted, which would be completely surprising to them and could cause a significant security hole.

The only safe way for us to do this without causing security problems is for the user to explicitly tell us that it's okay to trust xDS to tell us what credentials to use.  If they don't do that, we can't just assume it's okay to do so.  I understand that this does mean that applications need to make an additional change to use the xDS mTLS config, but we can't avoid it in this case.

I do not expect this same thing to come up for other features (e.g., retry options or LB policy).  It's a fairly security-specific constraint.

Reply:
@markdroth understood, thanks for the explanation. I will say if you flipped xds <-> grpc, it would all still make sense. Ultimately, we have two different sources of configuration for the same field, which is pretty tricky.

I can imagine a use case where I just want XDS for load balancing, and now I get a big security hole since the XDS server didn't send TLS configs. However, I can also see the inverse where someone wants essentially an Envoy replacement and is happy with the xDS being the complete source of truth. Have we explored exposing such information in the bootstrap configuration as well? One benefit of bootstrap vs code is the latter can be more easily changed I would imagine, and likely invisible to the user.

Reply:
I assume that the inverse case you're talking about is where a user was originally sending gRPC traffic through Envoy, so their client and server are both configured to use InsecureCreds (i.e., plaintext), and now they want to convert to proxyless.

In that situation, the first thing they would do is change the server side to use the new xDS-enabled server APIs that we're working on.  As part of that change, they would need to have their server listen on an external IP instead of localhost, and they would need to specify creds at the same time, so it should be fairly obvious that InsecureCreds is not the right thing anymore (although we should make sure to clearly document this).  And if they get the server side right, then if they change the client to use an "xds:" URI without using XdsCreds, then they'll immediately see the problem, because the client will be unable to connect to the server.

I don't think the bootstrap file solves this problem.  The fundamental issue here is that the user has told us what creds to use via the gRPC API, and I think we would severely violate their expectations if we do something different than what they told us due to some external source of data, regardless of whether that external source is the bootstrap file or the xDS server.